adapters/http-vertx-base/src/main/java/org/eclipse/hono/adapter/http/AbstractVertxBasedHttpProtocolAdapter.java

```diff
@@ -674,6 +674,11 @@ private void doUploadMessage(final RoutingContext ctx, final String tenant, fina
                             Optional.ofNullable(commandConsumerTracker.result()).ifPresent(consumer -> consumer.close(null));
                         });
                         ctx.response().end();
+                    } else {
+                        LOG.debug("failed to send http response for [{}] message from device [tenantId: {}, deviceId: {}]: response already closed",
+                                endpointName, tenant, deviceId);
+                        TracingHelper.logError(currentSpan, "failed to send HTTP response to device: response already closed");
+                        currentSpan.finish();
```

Comment:
in order to improve readability, I'd rather move this block to the top and check for `if (ctx.response().closed())` instead
WDYT?